### ISSUE TITLE
Any luck for PHP8 support?

### ISSUE THREAD

**Issue by youngvan** [iliaal/php_excel#264]
Greetings,

we would like to move to PHP8 when it will become available but we require php_excel module

any luck it will get php8 support?

as an option we can pay for that job

**Comment by Jan-E** [comment]
Windows or Linux?

**Comment by youngvan** [comment]
Linux

**Comment by Jan-E** [comment]
I cannot test it on Linux easily, but on Windows https://github.com/Jan-E/php_excel/commit/79ae81b369e714e3784cb6493eb1b89ac6cb81a9 was good enough. Some explanation:

1. The first change is to cast a long variable to time_t
2. At two places a semicolon was missing
3. LibXL is using 'int' for the arguments in xlSheetTableW(), so I changed the 'zend_long' into 'int'
4. I removed  TSRMLS_CC, which did not do anything at all in the PHP7 version anyway. See https://stackoverflow.com/a/49587574

These changes are good for PHP7 as well.

All of the changes to the original source of @iliaal are in the php7_with_pulls branch of https://github.com/Jan-E/php_excel
Proof of concept for Windows: https://www.apachelounge.com/viewtopic.php?t=6617

**Comment by youngvan** [comment]
Thanks a lot

here is error log i am getting with php8 rc4 and latest libxl release
https://pastebin.com/u2U3mey4

I can provide you linux environment so you can test

**Comment by youngvan** [comment]
here is one more output from clean install

https://pastebin.com/cEFSf5LT

**Comment by Jan-E** [comment]
I am getting more or less the same thing on Windows. See https://github.com/johmue/php_excel/commit/d13bc34e2195db7c15d52276f6b497b1d334b735#r44195623
On Windows they are merely warnings, no errors.

The 'too many parameters' can be solved by reverting https://github.com/Jan-E/php_excel/pull/1. Or maybe (as a quick fix) replacing the 4 instances of `PHP_EXCEL_SHEET_GET_BOOL_STATE_3831` by `PHP_EXCEL_SHEET_GET_BOOL_STATE`.

**Comment by Jan-E** [comment]
The 2 errors you are getting with `REGISTER_EXCEL_CLASS(Format,  format, excel_format_object_clone);` and `REGISTER_EXCEL_CLASS(Font,   font, excel_font_object_clone);` are reflected on Windows as `formal parameter 1 different from declaration`.

They do not happen on PHP 7, so this is a real difference between PHP 7 and PHP 8. As a very crude fix: replace `excel_format_object_clone` and `excel_font_object_clone` by NULL. Maybe that will solve the errors you are getting.

**Comment by youngvan** [comment]
Thanks for helping

here is much smaller log after all your suggestions

https://pastebin.com/Jkzs28H5

**Comment by youngvan** [comment]
as i see here is only one error:

/usr/include/libxl/libxl.h:374:30: error: too many arguments to function 'xlSheetIsFormulaA'
     #define xlSheetIsFormula xlSheetIsFormulaA
                              ^~~~~~~~~~~~~~~~~

**Comment by Jan-E** [comment]
I still had a copy of LibXL 3.8.4.0 somewhere. In thse sources include_c/SheetA.h contained
```
    XLAPI            int XLAPIENTRY xlSheetInsertColA(SheetHandle handle, int colFirst, int colLast, int updateNamedRanges);
    XLAPI            int XLAPIENTRY xlSheetInsertRowA(SheetHandle handle, int rowFirst, int rowLast, int updateNamedRanges);
    XLAPI            int XLAPIENTRY xlSheetRemoveColA(SheetHandle handle, int colFirst, int colLast, int updateNamedRanges);
    XLAPI            int XLAPIENTRY xlSheetRemoveRowA(SheetHandle handle, int rowFirst, int rowLast, int updateNamedRanges);
```
The latest sources did not have the third parameter for these 4 statements. But include_cpp/ISheetT.h does:
```
  virtual             bool XLAPIENTRY insertCol(int colFirst, int colLast, bool updateNamedRanges = true) = 0;
  virtual             bool XLAPIENTRY insertRow(int rowFirst, int rowLast, bool updateNamedRanges = true) = 0;
  virtual             bool XLAPIENTRY removeCol(int colFirst, int colLast, bool updateNamedRanges = true) = 0;
  virtual             bool XLAPIENTRY removeRow(int rowFirst, int rowLast, bool updateNamedRanges = true) = 0;
```
Seems a bit messy on the side of LibXL.

**Comment by Jan-E** [comment]
> /usr/include/libxl/libxl.h:374:30: error: too many arguments to function 'xlSheetIsFormulaA'
#define xlSheetIsFormula xlSheetIsFormulaA

Strange. Which version of LibXL are you using? What is the definition in SheerA.h? In my sources it is this:

```
   XLAPI            int XLAPIENTRY xlSheetIsFormulaA(SheetHandle handle, int row, int col);
```
That seems correct. The SheetHandle plus 2 aruments.

**Comment by youngvan** [comment]
I have tried 4 different version

right now with 3.8.4.0

definition is the same
    XLAPI            int XLAPIENTRY xlSheetIsFormulaA(SheetHandle handle, int row, int col);

error is the same

In file included from /usr/local/include/php/Zend/zend.h:27,
                 from /usr/local/include/php/main/php.h:31,
                 from /usr/src/php_excel/excel.c:26:
/usr/src/php_excel/excel.c: In function 'zim_ExcelSheet_isFormula':
/usr/include/libxl/libxl.h:374:30: error: too many arguments to function 'xlSheetIsFormulaA'
     #define xlSheetIsFormula xlSheetIsFormulaA
                              ^~~~~~~~~~~~~~~~~

**Comment by Jan-E** [comment]
I reverted back to LibXL 3.8.4.0 (for Win) as well. Complete build output:
```
Microsoft (R) Program Maintenance Utility Version 14.28.29333.0
Copyright (C) Microsoft Corporation.  All rights reserved.

        "cl.exe" /I "../win64build.vc16\include\libxl" /DHAVE_LIBXL_H=1 /D COMPILE_DL_EXCEL /D EXCEL_EXPORTS=1 /nologo /I . /I main /I Zend /I TSRM /I ext /D _WINDOWS /D WINDOWS=1 /D ZEND_WIN32=1 /D PHP_WIN32=1 /D WIN32 /D _MBCS /D _USE_MATH_DEFINES /FD /wd4996 /Qspectre /guard:cf /Zc:inline /Gw /Zc:__cplusplus /d2FuncCache1 /Zc:wchar_t /MP /Zi /LD /MD /Ox /D NDebug /D NDEBUG /D ZEND_WIN32_FORCE_INLINE /GF /D ZEND_DEBUG=0 /I "../win64build.vc16\include" /D FD_SETSIZE=1024 /DHAVE_ARGON2ID_HASH_RAW=1 /FoN:\php-sdk\php80dev\x64\Release\ext\excel\ /FpN:\php-sdk\php80dev\x64\Release\ext\excel\ /FRN:\php-sdk\php80dev\x64\Release\ext\excel\ /FdN:\php-sdk\php80dev\x64\Release\ext\excel\ /c ext\excel\excel.c
excel.c
ext\excel\excel.c(6830): warning C4028: formal parameter 1 different from declaration
ext\excel\excel.c(6831): warning C4028: formal parameter 1 different from declaration
Recreating build dirs
libiconv_a.lib(iconv1.obj) : MSIL .netmodule or module compiled with /GL found; restarting link with /LTCG; add /LTCG to the link command line to improve linker performance
   Creating library N:\php-sdk\php80dev\x64\Release\php8.lib and object N:\php-sdk\php80dev\x64\Release\php8.exp
Generating code
Finished generating code
        "link.exe" @"N:\php-sdk\php80dev\x64\Release\resp\EXCEL_GLOBAL_OBJS.txt" N:\php-sdk\php80dev\x64\Release\php8.lib libxl.lib kernel32.lib ole32.lib user32.lib advapi32.lib shell32.lib ws2_32.lib Dnsapi.lib psapi.lib bcrypt.lib argon2_a.lib N:\php-sdk\php80dev\x64\Release\php_excel.dll.res /out:N:\php-sdk\php80dev\x64\Release\php_excel.dll /dll /nologo /GUARD:CF /incremental:no /debug /opt:ref,icf /libpath:"../win64build.vc16\lib" /NXCOMPAT /DYNAMICBASE /libpath:"..\win64build.vc16\lib" /libpath:"..\win64build.vc16\lib"
   Creating library N:\php-sdk\php80dev\x64\Release\php_excel.lib and object N:\php-sdk\php80dev\x64\Release\php_excel.exp
EXT excel build complete
```
Note that the warnings about InsertRow et cetera are gone, but the different formal parameters are still there. No problem at all with IsFormula.

**Comment by Jan-E** [comment]
For the sake of argument I removed the `int col` in include_c/SheetA.h::
```
XLAPI int XLAPIENTRY xlSheetIsFormulaA(SheetHandle handle, int row,;
```
That resulted in the expected:
```
ext\excel\excel.c(2198): warning C4020: 'xlSheetIsFormulaA': too many actual parameters
ext\excel\excel.c(2654): warning C4020: 'xlSheetIsFormulaA': too many actual parameters
```
Ergo, the compiler is really looking at the number of argument in the `XLAPIENTRY xlSheetIsFormulaA`

**Comment by Jan-E** [comment]
Try replacing the `PHP_EXCEL_SHEET_GET_BOOL_STATE(IsFormula)` macro by the actual code. That should make it easier to debug:
```
diff --git a/excel.c b/excel.c
index 32cf5df..021695e 100644
--- a/excel.c
+++ b/excel.c
@@ -2651,7 +2651,14 @@ EXCEL_METHOD(Sheet, writeCol)
        Determine if the cell contains a formula */
 EXCEL_METHOD(Sheet, isFormula)
 {
-       PHP_EXCEL_SHEET_GET_BOOL_STATE(IsFormula)
+       SheetHandle sheet;
+       zval *object = getThis();
+       zend_long r, c;
+       if (zend_parse_parameters(ZEND_NUM_ARGS(), "ll", &r, &c) == FAILURE) {
+               RETURN_FALSE;
+       }
+       SHEET_FROM_OBJECT(sheet, object);
+       RETURN_BOOL(xlSheetIsFormula(sheet, r, c));
 }
 /* }}} */
```

**Comment by Jan-E** [comment]
Or even skip the `#define xlSheetIsFormula xlSheetIsFormulaA` and call `xlSheetIsFormulaA()` directly:
```
diff --git a/excel.c b/excel.c
index 32cf5df..655773c 100644
--- a/excel.c
+++ b/excel.c
@@ -2651,7 +2651,14 @@ EXCEL_METHOD(Sheet, writeCol)
        Determine if the cell contains a formula */
 EXCEL_METHOD(Sheet, isFormula)
 {
-       PHP_EXCEL_SHEET_GET_BOOL_STATE(IsFormula)
+       SheetHandle sheet;
+       zval *object = getThis();
+       zend_long r, c;
+       if (zend_parse_parameters(ZEND_NUM_ARGS(), "ll", &r, &c) == FAILURE) {
+               RETURN_FALSE;
+       }
+       SHEET_FROM_OBJECT(sheet, object);
+       RETURN_BOOL(xlSheetIsFormulaA(sheet, r, c));
 }
 /* }}} */

```

**Comment by Jan-E** [comment]
The starnge thing: in your first pastebin error log the IsFormula error was not there yet. What did change in between?

**Comment by youngvan** [comment]
thanks, was able to compile it now

the first pastebin was with latest libxl release and under docker

**Comment by Jan-E** [comment]
With LibXL 3.8.4.0? And with or without any further patches like https://github.com/iliaal/php_excel/issues/264#issuecomment-727579782 and/or https://github.com/iliaal/php_excel/issues/264#issuecomment-727594082 ?

If you patched any thing can you post it here?

**Comment by youngvan** [comment]
the first one was with 3.9.2

I want to replay all the process from the start to find minimum amount of changes and then I can share what i get

**Comment by Jan-E** [comment]
From the changelog:

Version 3.8.8 (2019-11-28)
--------------------------------------------------
- restored the updateNamedRanges parameter for Sheet::insertCol/insertRow/removeCol/removeRow C++ methods and added xlSheetInsertColAndKeepRanges / xlSheetInsertRowAndKeepRanges / xlSheetRemoveColAndKeepRanges / xlSheetRemoveRowAndKeepRanges C functions

In the C functions there is no third parameter anymore. But in the C++ functions we can see that the default for updateNamedRanges = true. The documentation says 
```
bool insertRow(int rowFirst, int rowLast, bool updateNamedRanges = true)
bool insertCol(int colFirst, int colLast, bool updateNamedRanges = true)
bool removeRow(int rowFirst, int rowLast, bool updateNamedRanges = true)
bool removeCol(int colFirst, int colLast, bool updateNamedRanges = true)
```
This should be reflected in the extension. Something like
1. Allow three parameters in Sheet::insertCol etc
2. If the third = true call xlSheetInsertColAndKeepRanges etc
3. If the third = false or 0 call xlSheetInsertCol etc (?)
4. If the third argument is omitted, use the default (true) ans call xlSheetInsertColAndKeepRanges etc

**Comment by Jan-E** [comment]
https://github.com/Jan-E/php_excel/commit/b0b1219cd28b57759d5592c182ccff11b90c9ba5 should fix the errors for a missing third argument. https://github.com/Jan-E/php_excel/tree/php7_with_pulls includes all patches.

**Comment by youngvan** [comment]
Yes, I can confirm that it works fine with latest libxl

Any chance it will be merged main repo one day? ;)

**Comment by Jan-E** [comment]
Fine that it works on Linux as well. Are you still getting this warning?
```
/root/php_excel/excel.c:6830:2: note: in expansion of macro ‘REGISTER_EXCEL_CLASS’
 6830 |  REGISTER_EXCEL_CLASS(Format,  format,   excel_format_object_clone);
      |  ^~~~~~~~~~~~~~~~~~~~
/root/php_excel/excel.c:85:47: warning: assignment to ‘zend_object_clone_obj_t’ {aka ‘struct _zend_object * (*)(struct _zend_object *)’} from incompatible pointer type ‘zend_object * (*)(zval *)’ {aka ‘struct _zend_object * (*)(struct _zval_struct *)’} [-Wincompatible-pointer-types]
```
@cmb69: this warning is generated through levels of macro's from this line:
https://github.com/Jan-E/php_excel/blob/php7_with_pulls/excel.c#L6910

The REGISTER_EXCEL_CLASS macro is here:
https://github.com/Jan-E/php_excel/blob/php7_with_pulls/excel.c#L75

And excel_format_object_clone is defined here:
https://github.com/Jan-E/php_excel/blob/php7_with_pulls/excel.c#L385

Have you got any idea where to look for the difference between PHP8 (with warning) and PHP7 (without warning)?

**Comment by cmb69** [comment]
The clone_obj handler expects a `zval*` in PHP 7, but a `zend_object*` in PHP 8; see https://github.com/php/php-src/blob/1727d96d0e1057ce140682b79d7f121dee0fc370/UPGRADING.INTERNALS#L236.

**Comment by Jan-E** [comment]
Thanks. I created a separate issue for this: https://github.com/iliaal/php_excel/issues/266

**Comment by Jan-E** [comment]
@youngvan Could you run the tests on https://github.com/Jan-E/php_excel/tree/php7_with_pulls (with 2 new patches)?

I noticed a cuple of things. some fail now with a 
`Fatal error: Uncaught ArgumentCountError`
Under PHP 7.x that wasn't a fatal error, but something like
`ExcelBook::packDateValues() expects exactly 6 parameters, 0 given`

There are some differences in the read/write tests (056.phpt). The first one fails with a 
`Warning: ExcelSheet::write(): Failed to write cell in row 5, column 2 with error 'format is null' in 056.php on line 10`

**Comment by youngvan** [comment]
I have played a lot with latest git pull and can confirm that it works fine under php7

**Comment by Jan-E** [comment]
TL;DR https://github.com/Jan-E/php_excel/tree/php7_with_pulls is compatible with PHP8 and the latest LibXL now.

**Comment by capsandiego** [comment]
@Jan-E Just compiled it with php-8.0.2 and libXL-3.8.5. It works like a charm. The world is a better place because of people like you.  Thank you so much!!

**Comment by capsandiego** [comment]
@Jan-E I fully understand that PHP8 is not yet fully supported but I see this error (which I think it's not difficult to fix):

```
$xlBook     = new \ExcelBook(null, null, $useXlsxFormat = true);
$xlBook->loadFile('test.xlsx');
$xlSheet   = $xlBook->getSheetByName("Sheet1", false);
$cellvalue = $xlSheet->read(0, 0, null, false);
```

It works perfectly with PHP7 but with PHP 8.0.2 on Linux I get this fatal error: **ExcelSheet::read(): Argument #3 ($format) cannot be passed by reference.**
I guess it's because of ```mixed &format``` in the php_excel.c ?
Thanks in advance!

**Comment by Jan-E** [comment]
Do you get a notice in PHP 7? See https://github.com/php/php-src/blob/1727d96d0e1057ce140682b79d7f121dee0fc370/UPGRADING#L142
> Some "Only variables should be passed by reference" notices have been
    converted to "Argument cannot be passed by reference" exception.

**Comment by johmue** [comment]
@capsandiego the issue is, that ˋread()ˋ uses $format to store an ExcelFormat object. Since you use a scalar/bool the ExcelFormat can not be stored in ˋ$formatˋ.

It‘s neither a bug nor needs to be fixed. It’s even documented. ...but I agree, it’s a very uncommon concept in PHP.

https://github.com/iliaal/php_excel/blob/936f353774916722c48b66afab4d4c4aff098ec9/docs/ExcelSheet.php#L841

**Comment by capsandiego** [comment]
Guys, thank you so much for your support. After your precious notes, I changed the script to this and it works fine with both PHP 7.4 and PHP 8.0.2. This was not a bug indeed.

```
error_reporting(E_ALL);
$xlBook        = new \ExcelBook(null, null, $useXlsxFormat = true);
$format        = new \ExcelFormat($xlBook);
$xlBook->loadFile('test.xlsx');
$xlSheet       = $xlBook->getSheetByName("Sheet1", false);
$cellvalue     = $xlSheet->read(0, 0, $format, $read_formula = false);
```
